src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedCallGraph.cs

```diff
@@ -0,0 +1,71 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+#nullable enable
+
+namespace ILCompiler.Dataflow
+{
+    sealed class CompilerGeneratedCallGraph
+    {
+        readonly Dictionary<TypeSystemEntity, HashSet<TypeSystemEntity>> callGraph;
+
+        public CompilerGeneratedCallGraph() => callGraph = new Dictionary<TypeSystemEntity, HashSet<TypeSystemEntity>>();
+
+        void TrackCallInternal(TypeSystemEntity fromMember, TypeSystemEntity toMember)
+        {
+            if (!callGraph.TryGetValue(fromMember, out HashSet<TypeSystemEntity>? toMembers))
+            {
+                toMembers = new HashSet<TypeSystemEntity>();
+                callGraph.Add(fromMember, toMembers);
+            }
+            toMembers.Add(toMember);
+        }
+
+        public void TrackCall(MethodDesc fromMethod, MethodDesc toMethod)
```

Comment:
We should add an assert that fromMethod and toMethod are in typical forms. I assume this won't do the right thing if things are instantiated.

(Same comment for the methods below.)